[CI:DOCS] GHA: Support testing build/sign workflows

cevich opened the pull request:
Neither `release` nor `workflow_dispatch` triggers may be tested inside a PR context.  The workflow steps always run from what's already committed to `main`.  Rather than waiting for a release to discover some unforeseen workflow problem, allow manual runs to optionally skip the release upload step (by default).

TODO:
- [ ] Test these changes on a throwaway `main` branch somewhere.
- [x] See if it's worth calling to a child-workflow to reduce steps duplication.

<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

In case you're only changing docs, make sure to prefix the pull-request title with "[CI:DOCS]". That will prevent functional tests from running and save time and energy.

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
None
```

ashley-cui reviewed (commented):
I think we should also change the if's in the build and sign stages to be `if: steps.check.outputs.build<arch> == 'true' || steps.set_actual.outputs.dryrun == 'true'`, so we guarantee the binaries are built even if the binaries already exist on a release

cevich commented:
@n1hility would you mind taking a look at my PowerShell butchery?  (I really don't know what I'm doing, just copy-pasting).  Two main things:

* Updated from the old/deprecated 'set-output' to the new file-based output setting method.
* Added a dry-run option to manual-runs to allow checking/verifying the build w/o publishing.

cevich commented:
/hold needs_testing

n1hility commented:
LGTM

@cevich the powershell changes looks correct to me, i don't see any issue, but would recommend doing a test release in case I missed anything.

cevich commented:
Thanks for taking a look @n1hility 

> but would recommend doing a test release in case I missed anything.

Ugg, yes, I'm dreading it, but do intend to test.  GHA makes this important step SOOOOOO painful, esp. when secrets are involved :cry:

I think I may try mocking out all the building and signing parts first, to check the dry-run logic is correct.

@n1hility and @ashley-cui are there commands I could add to the workflow that would verify the binary is correctly signed w/o attempting to install it?

cevich commented:
Thanks Jason, yeah it's the 'copy the secrets' part that's an extra PITA b/c there are so damn many of them (for both Mac and Win).  I think I'll just mock-out the actual build/signing stuff for testing.

Thanks for the test command, once this dry-run thing is in I can look into adding that and the equivalent for mac.

cevich commented:
Tests:
* Manual dry-run w/ non-existent version
  * [Windows](https://github.com/containers/automation_sandbox/actions/runs/5742533547)
  * [Mac](https://github.com/containers/automation_sandbox/actions/runs/5742530188)

ashley-cui commented:
For Macs: 

You can check if the .pkg is notarized by using 
`spctl -a -vvv -t install contrib/pkginstaller/out/podman-installer-macos-arm64.pkg`

To check if the binaries are correctly signed, you can use 
`codesign -dv --verbose=4 contrib/pkginstaller/out/packaging/root/podman/bin/podman`

cevich commented:
Thanks @ashley-cui I'll look into adding a check for that once this PR is done.  It will be a good excuse to try out the dry-run feature.

cevich commented:
Testing conclusion:

So in all cases, with a minor tweak and a check the dry-run changes operate properly.  However, IMHO there are a number of followup items that can go into separate PRs (I'm open for debate on all):

1. Add mandatory post-build/sign tests (commands in comments above) to verify the signed things are actually signed.  If not, still upload the workflow artifacts (to aid in debugging) but then fail the workflow with a helpful `::error::` message.
2. For the windows workflow.  It should check for the release asset `podman-remote-release-windows_amd64.zip`.  If it doesn't exist, fail the workflow with a helpful `::error::You didn't follow the podman RELEASE_PROCESS.md doc precisely enough.` (joking example) message.
3. For both mac and windows, for actual runs (not dry-runs), I think we can simplify/improve the workflows by simply always doing a `gh release upload --clobber $version <files>` at the end.  Reason being, we have the safety-net of a check (see item 1).  Alternatively, we could have a 'overwrite' input variable to control the behavior instead.  In the *unlikely* event of an accidental overwrite, there's a 90-day recovery window before prior workflow artifacts are deleted.

cevich commented:
@ashley-cui @n1hility PTAL when you have a chance, I think this is ready.  Though we should probably coordinate the followup work.  I'm a bit nervous to try and touch the windows/powershell scripts myself.

ashley-cui commented:
> 1. Add mandatory post-build/sign tests (commands in comments above) to verify the signed things are actually signed.  If not, still upload the workflow artifacts (to aid in debugging) but then fail the workflow with a helpful `::error::` message.

SGtm

> 3. For both mac and windows, for actual runs (not dry-runs), I think we can simplify/improve the workflows by simply always doing a `gh release upload --clobber $version <files>` at the end.  Reason being, we have the safety-net of a check (see item 1).  Alternatively, we could have a 'overwrite' input variable to control the behavior instead.  In the _unlikely_ event of an accidental overwrite, there's a 90-day recovery window before prior workflow artifacts are deleted.

The only possible mishap here is that we end up with a shasums file that includes old shasums, since we're not checking the contents of the shasums file, just appending to it. The mac script uploads the pkgs as an Actions artifact regardless of whether the GH upload succeeds or not. I'd rather err on the side of safety and not clobber, but I also don't feel super strongly about this.

ashley-cui reviewed (commented):
LGTM 😀

vrothberg reviewed (approved):
/lgtm

cevich commented:
> file that includes old shasums

It's easy to include that file into the workflow artifacts.

n1hility commented:
> 3. For both mac and windows, for actual runs (not dry-runs), I think we can simplify/improve the workflows by simply always doing a `gh release upload --clobber $version <files>` at the end.  Reason being, we have the safety-net of a check (see item 1).  Alternatively, we could have a 'overwrite' input variable to control the behavior instead.  In the _unlikely_ event of an accidental overwrite, there's a 90-day recovery window before prior workflow artifacts are deleted.

Hi @cevich! 

On the Windows side, I am a little worried about the possibility of ending up with multiple builds of the same version in the wild. The way the Windows installer architecture works is that every bulid gets a UUID, and this is how Windows recognizes the package is already installed (displaying options like uninstalling, reinstalling, and so on). Additionally, there is a contract where each version is supposed to map to one UUID, so it gets very confused when the same version is presented. So for example, if you were attempt to install a different build of the same version, the user would get a strange failure (likely version policy violation), and then have to know to uninstall it before installing the new build. 

Another issue is it might trigger suspicion with various malware tools, when they see checksum variations for the same version. 

While there could be circumstances where we want/need to do this (e.g.  broken build), ideally we would prevent accidental or CI dups causing it, limiting it to just an intentional action (e.g. explicit setting of the overwrite flag idea you list)

Thanks!

cevich commented:
Thanks so much Jason for the followup, I woudln't/couldn't have guessed any of that.  So the "existing installer" check is quite important then.  I wonder if the OSX environment is subject to the same/similar gotchas.  Since we don't (currently) have a check for that in place.  

@ashley-cui do you know?

ashley-cui commented:
As far as I'm aware, I don't think there's such a gotcha for the OSX installer, but I'm not an expert on this. I can double check.
